core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala

```diff
@@ -43,6 +42,18 @@ private[spark] class ApplicationInfo(
   @transient var endTime: Long = _
   @transient var appSource: ApplicationSource = _
 
+  // A cap on the number of executors this application can have at any given time.
+  // By default, this is infinite. Only after the first allocation request is issued
+  // by the application will this be set to a finite value.
+  @transient var executorLimit: Int = _
+
+  // A set of workers on which this application cannot launch executors.
+  // This is used to handle kill requests when `spark.executor.cores` is NOT set. In this mode,
+  // at most one executor from this application can be run on each worker. When an executor is
+  // killed, its worker is added to the blacklist to avoid having the master immediately schedule
+  // a new executor on the worker.
+  @transient var blacklistedWorkers: mutable.HashSet[String] = _
```

Comment:
I agree that in general we shouldn't have vars for mutable data structures, but this is marked as transient so unfortunately it needs to be a var. This just follows the pattern used in other variables in this class.

I believe the reason why these are marked as transient is because we don't want them to be set on the driver's side (e.g. the driver shouldn't try to blacklist executors). This doesn't have to be the case, but I wanted to minimize the changes in this patch.